in-progress/5040-native-merkle-trees-napi.md

```diff
@@ -0,0 +1,341 @@
+
+The correct way of running async operations is to wrap the code in an [AsyncWorker](https://github.com/nodejs/node-addon-api/blob/cc06369aa4dd29e585600b8b47839c1297df962d/doc/async_worker.md) so that the Nodejs runtime can track its execution properly:
+
+```cpp
+using tree_op_callback = std::function<bb::fr()>;
+class TreeOp : public AsyncWorker {
+  public:
+    TreeOp(Napi::Env env, Promise::Deferred& deferred, tree_op_callback& callback)
+        : AsyncWorker(env)
+        , _callback(callback)
+        , _deferred(deferred)
+        , _result(0)
+    {}
+
+    ~TreeOp() override = default;
+
+    void Execute() override
+    {
+        try {
+            _result = _callback();
+        } catch (const std::exception& e) {
+            SetError(e.what());
+        }
+    }
+
+    void OnOK() override { _deferred.Resolve(String::New(Env(), format(_result))); }
+    void OnError(const Napi::Error& e) override { _deferred.Reject(e.Value()); }
+
+  private:
+    tree_op_callback _callback;
+    Promise::Deferred _deferred;
+    bb::fr _result;
+}
+```
+
+`AsyncWorker.Queue` enqueues the execution of the worker at a later time on a thread managed by Node's libuv runtime.
+
+> [!IMPORTANT]
+> Inside `Execute()` code _must not_ access the JavaScript environment. This means everything that's needed to complete the operation _must_ be copied from the JS environment to memory owned by C++ before the task is queued up.
+> This also means that `Execute()` _can not_ create instances of `Napi::Value` since it does not have access to a `Napi::Env`.
+
+Once the async code finishes executing on the worker thread, one of the two event callbacks gets run on _the main NodeJS thread_. At this point the result of the async operation must be turned into a `Napi::Value` and returned back to the NodeJS code.
+
+### Memory limit
+
+NodeJS has a heap limit of about 4GB by default. This limit does not apply to the C++ module. The following code was used to allocate 40GB of RAM inside of a NodeJS process:
+
+```cpp
+// world_state_addon.hpp
+class WorldStateAddon : public Napi::ObjectWrap<WorldStateAddon> {
+  // ...
+  private:
+    std::vector<std::vector<std::byte>> _data;
+}
+
+// world_state_addon.cpp
+WorldStateAddon::WorldStateAddon(const Napi::CallbackInfo& info)
+    : ObjectWrap(info)
+{
+  // 40 * 1GB chunks
+  size_t chunks = 40;
+  for (size_t i = 0; i < chunks; i++) {
+    this->_data.emplace_back(1024 * 1024 * 1024);
+  }
+}
+```
+
+![40GB RAM](../images/node40gb.png)
+
+### Error handling
+
+Unhandled exceptions in the C++ code will crash the NodeJS process. Errors must be propagated correctly to the JS side if they can not be handled in C++.
+
+The C++ exceptions flag will be turned on at compile time so exception bubble naturally to the JS side. For async code errors should be returned by rejecting the associated promises.
```

Comment:
I recall @ludamad mentioned this was an issue when generating WASM. I assume it is not a problem in this context, right?